norse/torch/functional/stdp.py

```diff
@@ -1,60 +1,407 @@
+        dilation (int or tuple): convolutional dilation, defaults to 1 like nn.Conv2d
+    """
+
+    a_pre: torch.Tensor = torch.as_tensor(1.0)
+    a_post: torch.Tensor = torch.as_tensor(1.0)
+    tau_pre_inv: torch.Tensor = torch.as_tensor(1.0 / 100e-3)
+    tau_post_inv: torch.Tensor = torch.as_tensor(1.0 / 100e-3)
+    w_min: float = 0.0
+    w_max: float = 1.0
+    eta_plus: float = 1.0
+    eta_minus: float = 1.0
+    stride: Union[int, Tuple] = 1
+    padding: Union[int, Tuple] = 0
+    dilation: Union[int, Tuple] = 1
 
 
 class STDPState(NamedTuple):
-    """State of an event driven STDP.
+    """State of spike-timing-dependent plasticity (STDP)
 
     Parameters:
-        a_pre (torch.Tensor): presynaptic STDP state.
-        a_post (torch.Tensor): postsynaptic STDP state.
+        x (torch.Tensor): presynaptic spike trace
+        y (torch.Tensor): postsynaptic spike trace
     """
 
-    a_pre: torch.Tensor
-    a_post: torch.Tensor
+    x: torch.Tensor
+    y: torch.Tensor
 
 
-def A_plus_soft(w, w_max, eta_plus):
-    return (w_max - w) * eta_plus
+def _A_plus_soft(w, w_max, eta_plus):
+    """Soft upper bound for multiplicative STDP
 
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_max (float): upper bound on synaptic weight (should be > w_min)
+        eta_plus (float): contribution of pre-post activity to synaptic weight (should be >= 0)
+    """
+    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
+    return torch.nn.functional.relu(w_max - w) * eta_plus
+
+
+def _A_plus_hard(w, w_max, eta_plus):
+    """Hard upper bound for multiplicative STDP
 
-def A_plus_hard(w, w_max, eta_plus):
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_max (float): upper bound on synaptic weight (should be > w_min)
+        eta_plus (float): contribution of pre-post activity to synaptic weight (should be >= 0)
+    """
     return heaviside(w_max - w) * eta_plus
 
 
-def A_minus_soft(w, eta_minus):
-    return w * eta_minus
+def _A_minus_soft(w, w_min, eta_minus):
+    """Soft lower bound for multiplicative STDP
+
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_min (float): lower bound on synaptic weight (should be < w_max)
+        eta_minus (float): contribution of post-pre activity to synaptic weight (should be >= 0)
+    """
+    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
+    return torch.nn.functional.relu(w - w_min) * eta_minus
+
+
+def _A_minus_hard(w, w_min, eta_minus):
+    """Hard lower bound for multiplicative STDP
+
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_min (float): lower bound on synaptic weight (should be < w_max)
+        eta_minus (float): contribution of post-pre activity to synaptic weight (should be >= 0)
+    """
+    return heaviside(w - w_min) * eta_minus
+
+
+def linear_soft_multiplicative_stdp_step(
```

Comment:
I think this function and the function below might benefit from being a partial specialisations of a general function. We don't need to do it in this pull request. But it should happen soon.

Reply:
Good point. Would you mind creating an issue just to avoid loosing this gem of wisdom? :-)

Reply:
Good point I've done so here: #107.